chore: Clarify bytes metrics in component spec

binarylogic opened the pull request:
This PR updates the requirements for the `ComponentBytesReceived` and `ComponentBytesSent` events:

1. Renamed `ComponentBytesReceived` to `SourceNetworkBytesReceived`.
2. Renamed `ComponentBytesSent` to `SinkNetworkBytesSent`.
4. Clarified that the network bytes are best effort based on the source/sink capability. In other words, it should be the closest representation possible to the raw network bytes based on client limitations.
5. Clarified that the `ComponentEventsSent` and `ComponentEventsReceived` should emit a `byte_size` that reflects an estimated size for JSON encoding.

jszwedko commented on docs/specs/component.md:
```diff
@@ -103,101 +104,139 @@ events:
 
-*Sources* MUST emit a `ComponentBytesReceived` event immediately after receiving, decompressing
-and filtering bytes from the upstream source and before the creation of a Vector event.
+*Sources* MUST emit a `ComponentBytesReceived` event that represents the
+reception of *raw event bytes*. It MUST emit immediately after receiving
+bytes from the upstream source and before decompression, filtering, and the
+creation of Vector events.
```
To clarify, you mean for this metric to indicate the total number of bytes received on, for example, a TCP socket? This would include HTTP headers, gRPC envelopes, etc.

binarylogic replied:
No, _not_ the envelope/framing, just the raw bytes used to create events. For HTTP, this would be the HTTP body only (AKA the value of `Content-Length`). For a socket, this would be the bytes received minus the framing/delimiter.

binarylogic replied:
I clarify this in the requirements below.

jszwedko commented on docs/specs/component.md:
```diff
@@ -103,101 +104,139 @@ events:
+bytes sent downstream.
 
+- Emission
+  - MUST emit immediately after *successful* transmission the bytes.
+  - MUST emit before compression.
+  - MUST NOT emit for pull-based sinks since they do not send data. For
+    example, the `prometheus_exporter` sink MUST NOT emit this event.
```
Based on the naming, as a user, I'd expect this to be the inverse of `ComponentBytesReceived`. That is: it would indicate the raw bytes sent by the component over, for example, a TCP socket. This would mean both successful and unsuccessful requests. It would also mean pull based sinks would emit this metric.

binarylogic replied:
This does not include unsuccessful requests, it represents the successful delivery of bytes. This is an internal component metric, not a host-level network metric.

jszwedko reviewed (commented):
Agreed with publishing a separate metric for enterprise billing though I can see what you are saying about decoupling it in-case the billing strategy changes. I think `ComponentAcceptedBytes` is reasonable.

Do you mean for the `ComponentBytesReceived` and `ComponentBytesSent` events to indicate raw I/O now? I think there are some discrepancies below, if so. For example, pull-based sources will emit `ComponentBytesSent` for raw output for any requests they make. The `aws_ec2_metadata` transform will also emit raw I/O metrics. I might suggest separating out this change from the billing change to organize discussion around each.

Also, should we start keeping a changelog on the spec with this change? I think that'll make it easier to talk about the state of implementation of the spec.

binarylogic commented:
Talked with @jszwedko and simplified the requirements. I updated the description to summarize the changes.

tobz commented on docs/specs/component.md:
```diff
@@ -105,99 +108,121 @@ events:
-  - `http_path` - If relevant, the HTTP path, excluding query strings.
-  - `socket` - If relevant, the socket number that bytes were received from.
+  - `byte_size` - REQUIRED, number of raw bytes received.
+    - SHOULD be the total number of raw network bytes received *including*
+      framing (headers, delimiters, etc.). If If this is not possible due to a
+      client limitation, then the framed bytes MUST be used and the `framed` tag
+      MUST be set to `true`. 
```
So I find this wording a little confusing.

Based on the definition below of `framed`, I would expect `byte_size` to include the size of the entire request -- framing, headers, payload, all of it -- when `framed` is `true`... but here, you're saying that it's not possible to include the bytes related to framing/headers, then the "framed bytes" (huh? does this mean payload only?) must be used and `framed` must be set to `true`.

Overall, I think there needs to be a change to the wording to talk about the bytes as raw vs processed, _then_ refer to "framing" as all the additional bytes on top of the bytes we have to process aka decode, and then clear up the weird conditional statement about what `framed` means.

binarylogic replied:
I agree. See https://github.com/vectordotdev/vector/pull/12912#issuecomment-1206572535 for more info.

tobz commented on docs/specs/component.md:
```diff
@@ -105,99 +108,121 @@ events:
+      client limitation, then the framed bytes MUST be used and the `framed` tag
+      MUST be set to `true`. 
+  - `framed` - REQUIRED, `true` if the `byte_size` represents framed bytes,
+    `false` if not. See `bytes_size` for framing details.
```
`bytes_size` -> `byte_size`

tobz commented on docs/specs/component.md:
```diff
@@ -105,99 +108,121 @@ events:
+- Emission
+  - MUST emit immediately after transmission of raw network bytes regardless
+    if the transmission was successful or not.
+  - MUST emit *after* mutation of the bytes, such as decryption, decompression,
```
`decryption` -> `encryption`, `decompression` -> `compression`

tobz commented on docs/specs/component.md:
```diff
@@ -105,99 +108,121 @@ events:
 - Properties
-  - `count` - The count of Vector events.
-  - `byte_size` - The cumulative in-memory byte size of all events received.
+  - `byte_size` - REQUIRED, number of raw bytes sent.
+    - SHOULD be the total number of raw network bytes sent *including*
+      framing (headers, delimiters, etc.). If If this is not possible due to a
+      client limitation, then the framed bytes MUST be used and the `framed` tag
+      MUST be set to `true`. 
+  - `framed` - REQUIRED, `true` if the `byte_size` represents framed bytes,
```
Same comments as the ones I left on `ComponentsBytesReceived`.

tobz reviewed (changes requested):
I'm happy with a lot of the simplification happening here, but left some comments/requests on some of the language which feels a bit too ambiguous/hard to grok.

binarylogic commented on docs/specs/component.md:
```diff
@@ -55,13 +55,16 @@ follow the following guidelines.
+- MUST be a noun named after the protocol or service that the component
+  integrates with.
+- MAY be suffixed with the event type, `logs`, `metrics`, or `traces`
+  (e.g., `kubernetes_logs`, `apache_metrics`).
```
I think it's clear. This is an option if the source/sink is specific to the type.

binarylogic commented:
Ok, I pushed a big update that I think really clarifies the different between event bytes and network bytes. Here's what I changed:

1. Renamed `ComponentBytesReceived` to `SourceNetworkBytesReceived`.
2. Renamed `ComponentBytesSent` to `SinkNetworkBytesSent`.
3. Removed the framing nuance entirely. No `framed` tag, etc.
4. Clarified that the network bytes are best effort based on the source/sink capability. In other words, it should be the closest representation possible to the raw network bytes based on client limitations.
5. Clarified that the `ComponentEventsSent` and `ComponentEventsReceived` should emit a `byte_size` that reflects an estimated size for JSON encoding.

binarylogic commented on docs/specs/component.md:
```diff
@@ -103,104 +105,52 @@ events:
-  - `output` - For components that can use multiple outputs, the name of the
-    output that events were sent to. For events sent to the default output, this
-    value MUST be `_default`.
+  - `count` - REQUIRED, the number of Vector events.
+  - `byte_size` - REQUIRED, the estimated JSON byte size of all events sent.
```
Be more explicit this this is estimated and JSON

tobz replied:
Just making it match the wording/syntax of the `ComponentEventsReceived` event.

```suggestion
  - `count` - The count of Vector events.
  - `byte_size` - The estimated JSON byte size of all events sent.
```

binarylogic commented on docs/specs/component.md:
```diff
@@ -218,6 +168,66 @@ implement since errors are specific to the component.
 *All components* that can drop events MUST emit a `ComponentEventsDropped`
 event in accordance with the [EventsDropped event] requirements.
 
+#### SinkNetworkBytesSent
```
Cover billing with a post-processing size

jszwedko commented on docs/specs/component.md:
```diff
@@ -218,6 +168,66 @@ implement since errors are specific to the component.
+    example, the `prometheus_exporter` sink MUST NOT emit this event.
+- Properties
+  - `byte_size` - REQUIRED, number of raw network bytes sent after processing.
+    - SHUOLD be the closest representation possible of raw network bytes based
```
```suggestion
    - SHOULD be the closest representation possible of raw network bytes based
```

jszwedko commented on docs/specs/component.md:
```diff
@@ -218,6 +168,66 @@ implement since errors are specific to the component.
+    if the transmission was successful or not.
+  - MUST emit *after* processing of the bytes (encryption, compression,
+    filtering, etc.)
+  - MUST NOT emit for pull-based sinks since they do not send data. For
+    example, the `prometheus_exporter` sink MUST NOT emit this event.
```
Given the change here to be oriented on network bytes, do we still want to skip pull-based sinks?

binarylogic replied:
Nope, good call, updated.

jszwedko commented on docs/specs/component.md:
```diff
@@ -218,6 +168,66 @@ implement since errors are specific to the component.
+      on the sink's capabilities. For example, if the sink uses a HTTP
+      client that does not provide access to the total request byte size, then
+      the sink should use the byte size of the payload/body.
+  - `url` - REQUIRED, the URL of the downstream destination.
+    - For HTTP, MUST not include the query string, only the host and path.
+      (i.e., `http(s)://<host>:<port>/<path>`)
```
This turns out to not be a great idea for sinks like the AWS S3 sink where the object key is part of the URL.

My suggestion would be that we drop `url` entirely until someone asks for it. I think the only sink where it'd be useful to users is the `http` sink, but even then I'm not sure.

jszwedko commented on docs/specs/component.md:
```diff
@@ -218,6 +168,66 @@ implement since errors are specific to the component.
+  - `url` - REQUIRED, the URL of the downstream destination.
+    - For HTTP, MUST not include the query string, only the host and path.
+      (i.e., `http(s)://<host>:<port>/<path>`)
+    - For sockets, the full socket URL (i.e., `tcp://<address>:<port>`)
+    - For files, MUST be the full file url (i.e., `file://<path>`)
```
Also here, I suggest we drop these until someone asks for it. I can suggest these changes in a follow-up PR if we prefer, since they are moved. I just figured since this is a new event it is a good time to make these additional broad changes.

jszwedko commented on docs/specs/component.md:
```diff
@@ -218,6 +168,66 @@ implement since errors are specific to the component.
+- Properties
+  - `byte_size` - REQUIRED, number of raw network bytes received before
+    processing (decryption, decompression, filtering, etc.).
+    - SHUOLD be the closest representation possible of raw network bytes based
```
```suggestion
    - SHOULD be the closest representation possible of raw network bytes based
```

jszwedko commented on docs/specs/component.md:
```diff
@@ -218,6 +168,66 @@ implement since errors are specific to the component.
+      on the source's capabilities. For example, if the source uses a HTTP
+      client that only provides access to the request body, then the raw
+      request body bytes should be used.
+  - `url` - REQUIRED, the URL of the upstream source.
+    - For HTTP, MUST not include the query string, only the host and path.
+      (i.e., `http://<host>:<port>/<path>`)
+    - For sockets, the full socket URL (i.e., `tcp://<address>:<port>`)
+    - For files, MUST be the full file url (i.e., `file://<path>`)
```
Same here, I suggest we drop these tags and simplify. I think we went a little overboard on the initial metric definition.

jszwedko reviewed (commented):
👍 to these changes in general. I left some comments below inline.

For the billing metric, is my reading correct that we would now use `component_received_event_bytes_total` on sources?

binarylogic commented:
> For the billing metric, is my reading correct that we would now use component_received_event_bytes_total on sources?

Assuming this is an accurate representation of the JSON encoding, I don't see why not. My only concern is enriching within the source and how customers would feel about that. Regardless, I don't want to encourage customers to use an inferior protocol because it's cheaper. We should try to normalize the bytes we're charging post-process.

jszwedko commented on docs/specs/component.md:
```diff
@@ -103,104 +105,52 @@ events:
-    defined properties as metric tags.
-  - MUST increment the `component_sent_event_bytes_total` counter by the event's byte size
-    in JSON representation.
+  - MUST increment the `component_sent_events_total` counter by the defined
+    value with the defined properties as metric tags.
+  - MUST increment the `component_sent_event_bytes_total` counter by the event's
+    byte size in JSON representation.
```
```suggestion
  - MUST increment the `component_sent_events_total` counter by the defined
    `quantity` property with the other properties as metric tags.
  - MUST increment the `component_sent_event_bytes_total` counter by the
    defined `byte_size` property with the other properties as metric tags.
```

Just making this match up with the received event language.

jszwedko reviewed (approved):
Looks good to me! I do think we should start tracking revisions to this document as a changelog in the doc itself to be able to easily refer to versions of the doc, but we can start that with any future ones.

tobz commented on docs/specs/component.md:
```diff
@@ -218,6 +168,57 @@ implement since errors are specific to the component.
+*raw network bytes*.
+
+- Emission
+  - MUST emit immediately after egress of raw network bytes regardless
+    if the transmission was successful or not.
```
```suggestion
  - MUST emit immediately after egress of raw network bytes regardless
    of whether the transmission was successful or not.
```

tobz commented on docs/specs/component.md:
```diff
@@ -218,6 +168,57 @@ implement since errors are specific to the component.
+  - MUST emit *after* processing of the bytes (encryption, compression,
+    filtering, etc.)
+- Properties
+  - `byte_size` - REQUIRED, number of raw network bytes sent after processing.
```
```suggestion
  - `byte_size` - The number of raw network bytes sent after processing.
```

tobz commented on docs/specs/component.md:
```diff
@@ -218,6 +168,57 @@ implement since errors are specific to the component.
+    filtering, etc.).
+    - This includes pull-based sources that issue requests to ingest bytes.
+- Properties
+  - `byte_size` - REQUIRED, number of raw network bytes received before
```
```suggestion
  - `byte_size` - The number of raw network bytes received before
```

tobz reviewed (approved):
Overall, this looks good and makes sense to me.

I left some suggestions for improving clarity, but they're non-blocking.

JeanMertz reviewed (approved):
All of this makes sense to me 👍
